### ISSUE TITLE
Plugin Auto system failing

### ISSUE THREAD

**Issue by rgleason** [Rasbats/shipdriver_pi#364]
I can simply copy the CS deployed metadata to plugins,
or I can get the automatic going, but that is failing.
Does anyone have a suggestion to fix this?

```
fcgle@Hesper MINGW64 ~/source/vdr_pi (master)
$ python ./ci/new-credentials
Traceback (most recent call last):
  File "C:\Users\fcgle\source\vdr_pi\ci\new-credentials", line 23, in <module>
    from cryptography.fernet import Fernet, InvalidToken
ModuleNotFoundError: No module named 'cryptography'

```

**Comment by leamas** [comment]
Anything you do in the build chain requires that the build dependencies are installed: https://opencpn-manuals.github.io/main/AlternativeWorkflow/Local-Build.html

**Comment by rgleason** [comment]
I have all the scripts installed, and paths are all correct,  Is this another problem with script permissions?

Is there a way to run  update-templates now just to take care of all this Cxxx?

**Comment by leamas** [comment]
> I have all the scripts installed, and paths are all correct, Is this another problem with script permissions

No, the problem is that you have not run build_win.bat which installs the build dependencies.

**Comment by rgleason** [comment]
Ok I have run it now as admin. I hope it does not screw up my other installations.
It looks like you're creating a cache with wxwidgets-3.1.2  ...
[win-deps.bat.output.txt](https://github.com/Rasbats/shipdriver_pi/files/7738209/win-deps.bat.output.txt)

**Comment by rgleason** [comment]
I think I should try to configure the auto metadate push to plugins again.

I tagged for beta   tag v1.3.13.beta and pushed tags it worked and built everything {I think).
I pushed a commit to remove rasbats enc and pub.
I tagged for prod  tag v1.3.13  and pushed tags and they all worked except drone and appveyor and they failed.
do you think those are related?

appveyor failed
```
CUSTOMBUILD : CMake error : Could not open file for write in copy operation C:/project/opencpn/plugin_src/build/VDR-v1.3.13 [C:\project\opencpn\plugin_src\build\tarball-build.vcxproj]
  v1.3.13.beta-msvc-10.0.14393.xml.in.tmp
CUSTOMBUILD : CMake error : : System Error: Invalid argument [C:\project\opencpn\plugin_src\build\tarball-build.vcxproj]
  CMake Error at CMakeLists.txt:107 (configure_file):
    configure_file Problem configuring file
  -- Configuring incomplete, errors occurred!
  See also "C:/project/opencpn/plugin_src/build/CMakeFiles/CMakeOutput.log".
Command exited with code 1
```

drone failed too.
https://cloud.drone.io/rgleason/vdr_pi/43/1/2

**Comment by rgleason** [comment]
The manual should state clearly these steps from https://opencpn-manuals.github.io/main/AlternativeWorkflow/Local-Build.html#_the_local_build_environment - it is too buried:
1. Modify  scripts/local-build.rc  with the necessary personal information and keys.
2. Copy scripts/local-build.rc to %appsdata%  under "roaming"

https://opencpn-manuals.github.io/main/AlternativeWorkflow/Local-Build.html#_the_configuration_file
A sample configuration file is distributed in scripts/local-build.rc. Instructions for modifying and installation is included in the sample. Note that all other tools depends on this file. 
--Where is this?  It is pretty important I think and should be more accessible and in the text.

**Comment by rgleason** [comment]
https://opencpn-manuals.github.io/main/AlternativeWorkflow/InstallConfigure/Catalog-Github-Integration.html
```
These items should already have been carried out:

   1. Run python ci/new-credentials in the TideFinder repo. The script will ask for the ssh url to your plugins fork repo and a secret password. The ssh url, something like git@github.com:Rasbats/plugins.git, can be found in the Github UI:
```

"These should already have been carried out"   - SO WHERE in the PROCESS!

Also I ran the script  `python ci/new-credentials` in the vdr_pi local repo
and it did NOT PROMPT for a password and it is right now stuck and not completed and I have no idea of what to do.
There are new rgleason.pub and rgleason files in build-deps. 

Also I am trying to get this working despite the lack of password, and am asked at the same time to insert , is it the PUBLIC key???  in  a new rgleason/plugins  Deploy Key.   Then what happens? Will my open bash prompt with the credentials automatically close (without a Password?)   This should be broken into more specific steps.

**Comment by rgleason** [comment]
Having completed saving the new deploy key in rgleason/plugins  as vdr deply, the open bash cli with the unfinished or incomplete new-credentials process, did not close and I am stuck here.

**Comment by rgleason** [comment]
Tried   ./ci/new-credentials; exit
but it did not work.  So I just closed it despite open processes.

**Comment by rgleason** [comment]
rgleason and rgleason.pub are still present.

**Comment by rgleason** [comment]
I was wondering about deployment of both keys,  but deployed them. Shortly afterwards I got this notice.

```
We noticed that a valid SSH private key of yours was committed to a GitHub repository. This key is configured as a deploy key for the rgleason/plugins repository. Disclosing a valid SSH private key would allow other people to interact with this repository, potentially altering data.

As a precautionary measure, we have unverified the SSH key. You should generate a new SSH key and add it to the repository. We recommend you review your security log to ensure that no malicious activity has occurred: https://docs.github.com/articles/reviewing-your-security-log/

The commit in question is at https://github.com/rgleason/vdr_pi/blob/8dc3c1399b4d791084b2551d4a9832a7327ad401/build-deps/rgleason

Please feel free to contact us at https://github.com/contact if you have any questions or concerns.

Thanks,
GitHub.com
```

Thank you github.

**Comment by rgleason** [comment]
From the instructions:

```
Dont forget the password…​

The script will create an encrypted private + a public key in the TideFinder build-deps directory. Commit and push these to Github:

 git add build-deps/*.enc build-deps/*.pub

 git commit -m "Add git deployment key."

 git push origin master:master

Register the new public key (displayed by new-credentials) as a deployment key with write permissions (Allow write access) on your github OpenCPN/plugins fork repo (in Settings | Deploy keys):
```

1. There is no password, the script does not prompt for one.
2. Do NOT commit and push these to GITHUB!    - so what should we do with the private key?
3. How to get out of the script satisfactorily?
4. Unfortunately this is not working yet.

**Comment by rgleason** [comment]
I am waiting with a new public and private key but will not push them until this is straightened out.  I have removed the old key and added a new one.

All of the ci/ scripts built   see https://github.com/rgleason/vdr_pi/commits/master
and seem to have deployed to PROD https://cloudsmith.io/~opencpn/repos/vdr-prod/packages/
so I am just going to use Jon's download_sml_bash.sh script to upload them to plugins for this go round.

**Comment by rgleason** [comment]
After the message from Github, I was offered a free opensource account with gitguardian which I accepted..
There are some risks shown for some of the repositories,  some from March 17  regarding development.

**Comment by rgleason** [comment]
vdr v1.3.14 is pushed to plugins now.

**Comment by leamas** [comment]
WARNING: YOU HAVE PUSHED YOUR PRIVATE KEY INTO A PUBLIC REPOSITORY!!!

This is a major disaster. The key is compromised. Steps to take
  - Remove both rgleason and rgleason.pub from the repo, and push
  - Remove the deployment key you added  in the Github UI.

That should be it.

> We noticed that a valid SSH private key of yours was committed to a GitHub repository.

That was a really helpful piece of automation on github...

**Comment by leamas** [comment]
> "These should already have been carried out" - SO WHERE in the PROCESS!

Agreed, this is a documentation error. I have made a PR. Just disregard this paragraph, and continue with next step which has become nr 1 but really is 3.

**Comment by leamas** [comment]
> Also I ran the script python ci/new-credentials in the vdr_pi local repo
and it did NOT PROMPT for a password and it is right now stuck[...}

You cannot proceed until you have resolved this.   The attempt to proceed despite not being able to run _new_credentials_ is what caused your major security violation.

Under the first number 1 point "Preparations" you can find the steps you need to take to avoid that _new_credentials_ get stuck like you describe.

Note that both code and documentation has been tested, lately by Håkan, so there should be no major problems. That is not to say there are no coŕner-cases.

**Comment by rgleason** [comment]
The cause of the security violation was the fact that for some reason the script did not ask for a password!
Not everybody is going to know that there will be security violation, if the passphrase is not prompted!

What causes the pass phrase prompt not to appear?

**Comment by leamas** [comment]
So, you basically cannot run the script. Is this still so? 

That is, if you follow first step 1 (preparations) and then step  3 (run new-credentials) , does this work for you? It should definitely prompt you for url and password. Does it?

**Comment by rgleason** [comment]
1. Removed deployment plugin key in github
2. Removed rgleason.pub and rgleason from the repo
3. Removed /local-build.rc from %appsdata% under "roaming"
4. Removed the keys from git history  - git checked out two commits back and then --force pushed to origin master
5. Removed logbookkonni and vdr from plugins.

**Comment by rgleason** [comment]
Repeat:  
 [GitGuardian](https://github.com/GitGuardian) https://github.com/GitGuardian shows that there are many insecure API keys used in plugins that I have forked.  Many of these keys are for Cloudsmith and should be fixed as soon as possible.  Use gitguardian to find them.

**Comment by rgleason** [comment]
```
fcgle@Hesper MINGW64 ~/source/plugins (master)
$ git config --get remote.origin.url
git@github.com:rgleason/plugins.git

```
Then


```
C:\Users\fcgle\source\vdr_pi>

C:\Users\fcgle\source\vdr_pi>python ci/new-credentials
python ci/new-credentials
Enter ssh URL for ocpn-plugins.xml clone: git@github.com:rgleason/plugins.git
Generating public/private rsa key pair.
Your identification has been saved in build-deps/rgleason
Your public key has been saved in build-deps/rgleason.pub
The key fingerprint is:
SHA256:xxxxxxxxxxxxxfcgle@Hesper
The key's randomart image is:
+---[RSA 3072]----+
|  .=*o.  .+      |
----
---
|*  = X + .       |
|Bo. * * S E      |
|== o o           |
|=.o .            |
|++ .             |
|+.o              |
+----[SHA256]-----+

```

NO PROMPT

**Comment by rgleason** [comment]
Regarding Item 1  Python is working and I have an upstream for plugins.
```
fcgle@Hesper MINGW64 ~/source/plugins (master)
$ git branch -a
  auto
* master
  remotes/origin/Alpha
  remotes/origin/Beta
  remotes/origin/HEAD -> origin/master
  remotes/origin/auto
  remotes/origin/master
  remotes/upstream/Alpha
  remotes/upstream/Beta
  remotes/upstream/Experimental
  remotes/upstream/master


```

**Comment by leamas** [comment]
Have you defined   the python alias as described in step 1 Preparation before doing this?

**Comment by rgleason** [comment]
I did it last time, not this time.  The keys are found in build-deps but I am deleting them as I don't want the pushed up!

**Comment by leamas** [comment]
Correct, don't push fthe keys. 
> I did it last time, not this time
¨
It must be done every time. Look into the stackoverflow  question for a permanent solution.

The alias is required to avoid exactly the situation you describe

> . The keys are found in build-deps

No. If you look into step 5 you can see that the keys you should push is rgleason.pub and rgleason.enc. However, the .enc file does not exist since you have not been able to run the script.

> I am deleting them as I don't want the pushed up!

Correct action until you are able to run the script, agreed.-

**Comment by rgleason** [comment]
The script also did not complete, so I had to just close it and delete the lock.

**Comment by leamas** [comment]
This is actually as expected without an alias. See the stackoverflow question for the background.

**Comment by leamas** [comment]
I have started my not that much used Windows VM, cloned vdr_pi and tried running ` python ci/new-credentials`

  - Trying without an alias, I get the error you describe
  - If I define the alias, everything works as expected.

**Comment by rgleason** [comment]
Yes, this worked:
From bash prompt in github.com/<user>/plugins/master
```
$ git config --get remote.origin.url
git@github.com:rgleason/plugins.git

$ alias python='winpty python.exe'
$ python --version
Python 3.10.0

Then change bash prompt to the vdr_pi folder
fcgle@Hesper MINGW64 ~/source/vdr_pi (master)
$ python ci/new-credentials
Enter ssh URL for ocpn-plugins.xml clone: git@github.com:rgleason/plugins.git
Generating public/private rsa key pair.
Your identification has been saved in build-deps/rgleason
Your public key has been saved in build-deps/rgleason.pub
The key fingerprint is:
Then shows randomart image:
New key build-deps/rgleason  successfully created

Enter password for encrypting new key:
Repeat password:

New ssh key created and encrypted.Todo:
  - Commit the new ssh keys build-deps/rgleason.pub and build-deps/rgleason.enc.
  - Add the ssh url as GIT_REPO environment variable in builder.
  - Add password as GIT_KEY_PASSWORD environment variable in builder.
  - Add the public key build-deps/rgleason.pub below to the github account:
```

**Comment by rgleason** [comment]
It is still not building locally. Is that because of the keys are missing?
```
  -- Selecting Windows SDK version 10.0.17763.0 to target Windows 10.0.19042.
  -- Selected upload repository: o p e n c p n / v d r - b e t a 
  -- Install Prefix: C:/Users/fcgle/source/vdr_pi/build/app/files
CUSTOMBUILD : CMake error : Could not open file for write in copy operation C:/Users/fcgle/source/vdr_pi/build/VDR-v1.3.13 [C:\Users\fcgle\source\vdr_pi\build\tarball-build.vcxproj]
  v1.3.13.beta-msvc-10.0.19042.xml.in.tmp
CUSTOMBUILD : CMake error : : System Error: Invalid argument [C:\Users\fcgle\source\vdr_pi\build\tarball-build.vcxproj]
  CMake Error at CMakeLists.txt:107 (configure_file):
    configure_file Problem configuring file
  
```
  
  -- Configuring incomplete, errors occurred!

**Comment by leamas** [comment]
Please take everything in the right order. Have you completed all the steps in https://opencpn-manuals.github.io/main/AlternativeWorkflow/InstallConfigure/Catalog-Github-Integration.html

**Comment by rgleason** [comment]
Yes, I will do so, but what is necessary to check the build locally? 
Do I really need to enable all the Automatic push to Plugins stuff?

**Comment by leamas** [comment]
We just cannot make ell these efforts to get half-way through the procedure in vain. You need to push your keys and register the deployement key while they are available

**Comment by leamas** [comment]
BTW: What you push now is **not** your private key, it's your **encrypted** private key which is fine -- nobody can use it unless they know the  password you gave to _new-credentials_. Which you will keep safe.

**Comment by rgleason** [comment]
Yes, I am doing it.  Should I use these keys for Logbook too, or not?

**Comment by leamas** [comment]
You can use them everywhere.

**Comment by rgleason** [comment]
So I can use the same keys for Logbook.  Thanks.

**Comment by rgleason** [comment]
https://github.com/Rasbats/shipdriver_pi/issues/364

Enter this into all the "Builders"  for each plugin  (vdr_pi  and logbookkonni_pi)

```
    CLOUDSMITH_API_KEY    comment out
    GIT_REPO  git@github.com:rgleason/plugins.git
    GIT_KEY_PASSWORD   comment out

```
           vdr   logbook
Appveyor - done done
Circleci  - done done 
Github Actions  - request
Drone.io - done done
Travis - done done     Still shutdown due to credit problem

**Comment by rgleason** [comment]
Have requested  Git Actions access https://github.com/leamas/shipdriver_pi/issues/3

Should I configure Custom Cloudsmith Repos?
https://opencpn-manuals.github.io/main/AlternativeWorkflow/InstallConfigure/BuilderEnv.html#_custom_cloudsmith_repositories

What next?

**Comment by rgleason** [comment]
For Local builds, the https://opencpn-manuals.github.io/main/AlternativeWorkflow/Local-Build.html#_overview 
the change of this file is really glossed over   C:\Users\fcgle\source\vdr_pi\scripts\local-build.rc
Is it necessary or not?  The document does not indicate where this file is located either?

```
export CLOUDSMITH_API_KEY=39afdeadbeefxxerrnotevenhexdigitsher
export GIT_KEY_PASSWORD=MySecretPassword
export GIT_REPO=git@github.com:leamas/plugins.git

# These variables are only required when using upload directories others
# than those defined in Plugin.cmake. To be effective they must be available
# during the build

export CLOUDSMITH_STABLE_REPO=alec-leamas/opencpn-plugins-stable
export CLOUDSMITH_UNSTABLE_REPO=alec-leamas/opencpn-plugins-unstable
export CLOUDSMITH_BETA_REPO=alec-leamas/opencpn-plugins-stable
 
```

**Comment by rgleason** [comment]
I was just able to build locally without making any changes to local-build.rc.
Do I need to change it or not?  What is it for?
Also do I need to put a copy in %appsdata%/roaming ?  If so, why?

**Comment by rgleason** [comment]
https://opencpn-manuals.github.io/main/AlternativeWorkflow/Local-Build.html#_testing_the_build
It looks like completing the script is needed. and should be copied to %appsdata%  to push to plugins under certain circumstances. (what are they?)

**Comment by rgleason** [comment]
I have completed that script revision of local-build.rc

However there are differences from what is shown in plugin.cmake  shown below:

```
set(OCPN_TEST_REPO
    "opencpn/vdr-alpha"
    CACHE STRING "Default repository for untagged builds"
)
set(OCPN_BETA_REPO
    "opencpn/vdr-beta"
    CACHE STRING "Default repository for tagged builds matching 'beta'"
)
set(OCPN_RELEASE_REPO
    "opencpn/vdr-prod"
    CACHE STRING "Default repository for tagged builds not matching 'beta'"
)

```

Shouldn't this be corrected?

**Comment by rgleason** [comment]
I changed local-build.rc and saved it to %appdata%
Then commited those changes and made a beta tag and pushed.
https://github.com/rgleason/vdr_pi/commits/master

**Comment by rgleason** [comment]
The local version seems to work.

**Comment by rgleason** [comment]
Pushed [version works](https://github.com/rgleason/vdr_pi/commits/master) and went to beta as expected. 
https://cloudsmith.io/~opencpn/repos/vdr-beta/packages/


I think this [manual push to plugins  is a very helpful page](https://opencpn-manuals.github.io/main/AlternativeWorkflow/InstallConfigure/Catalog-Github-Integration.html#_making_the_pull_request_pr_to_opencpnplugins) and it should be kept, but with a note saying it is manual push and superceded by the automatic push to plugins [provide a link to the page]

**Comment by rgleason** [comment]
I think I can almost close this now. I made a new tag v1.3.15 and pushed it and it landed in the opencpn/vdr-prod
https://cloudsmith.io/~opencpn/repos/vdr-prod/packages/

Will it automatically deploy to Plugins?

There was a message in my remote plugins repos about a push.  I then made a PR and let it check.
I hope I did not do it too early and miss some of the necessary builds!!!! How would I know or find out?

https://github.com/OpenCPN/plugins/pull/560

**Comment by rgleason** [comment]
Darn, errors

https://github.com/OpenCPN/plugins/runs/4577289931?check_suite_focus=true

**Comment by rgleason** [comment]
It looks like there were two sets of commits with different SHA which may have caused the rejection.

**Comment by leamas** [comment]
When I clone your repo I cannot build anything form it, it just ignores the source files and builds an empty tarball, at best.

Seems that you have updated from master branch. This is seldom a great idea, especially when not on stable ground otherwise.

The last released version is 3.0.0 which is visible in the Github UI as the sd3.0.0 tag. Downgrading to this seems to resolve things:

    $  ./update-templates sd3.0.0
    $ git checkout sd3.0.0 CMakeLists.tst

The examples in the documentation should probably use a tag rather than defaulting to shipdriver/master.

**Comment by leamas** [comment]
> Do I need to change it or not? What is it for?

A plain build works without this file. However, if you want to make an upload, a git-push  or using scripts/local-build 
which basically does all of it the file is required.

> so do I need to put a copy in %appsdata%/roaming ? If so, why?

Because you obviously cannot put your secrets in file which is pushed to a public git repo. The file in the repo is just a template, the tools which needs it does not look for it but rather what's available in %appsdata%/roaming (on Windows).
